### PR TITLE
BUX-325 [bux-wallet-bk] Generic error handling

### PR DESCRIPTION
**Pull request by ac4ch**
…x client operations

# Pull Request Checklist

- [ ] 📖 I created my PR using provided  : [CODE_STANDARDS](https://github.com/bitcoin-sv/bux-wallet-backend/blob/main/.github/CODE_STANDARDS.md)
- [ ] 📖 I have read the short Code of Conduct: [CODE_OF_CONDUCT](https://github.com/bitcoin-sv/bux-wallet-backend/blob/main/.github/CODE_OF_CONDUCT.md)
- [ ] 🏠 I tested my changes locally.
- [ ] ✅ I have provided tests for my changes.
- [ ] 📝 I have used conventional commits.
- [ ] 📗 I have updated any related documentation.
- [ ] 💾 PR was issued based on the Github or Jira issue.

### CODE REVIEW

**Review by dorzepowski — APPROVED** [review]

File: domain/users/users_service.go [hunk 1]
```diff
@@ -6,27 +6,63 @@ import (
...
+	message string
+}
+
+func (e *XPubError) Error() string {
+	return e.message
+}
+
 // ErrInvalidCredentials is throwing when invalid credentials were used.
-var ErrInvalidCredentials = errors.New("invalid credentials")
+var ErrInvalidCredentials = &CredentialsError{"invalid credentials"}
```
Comment (arkadiuszos4chain) [inline domain/users/users_service.go:62]: What are the benefits of this custom structure?

File: transports/http/endpoints/api/users/endpoints.go [hunk 2]
```diff
@@ -71,11 +69,16 @@ func (h *handler) register(c *gin.Context) {
...
-			return
+		switch err.(type) {
+		case *users.UserError:
+			c.JSON(http.StatusBadRequest, fmt.Sprintf("Error creating user: %v", err))
+		case *users.PaymailError:
+			c.JSON(http.StatusBadRequest, fmt.Sprintf("Error registering Paymail: %v", err))
+		case *users.XPubError:
+			c.JSON(http.StatusBadRequest, fmt.Sprintf("Error registering XPub: %v", err))
+		default:
+			c.JSON(http.StatusInternalServerError, "Something went wrong when creating new user. Please try again later.")
```
Comment (Nazarii-4chain) [inline transports/http/endpoints/api/users/endpoints.go:80]: Good improvements 👍

**Review by Nazarii-4chain — APPROVED** [review]
LGTM

File: transports/http/endpoints/api/users/endpoints.go [hunk 3]
```diff
@@ -71,11 +69,16 @@ func (h *handler) register(c *gin.Context) {
 
 	// Check if user with this email already exists or there is another error
 	if err != nil {
-		if err == users.ErrUserAlreadyExists {
-			c.JSON(http.StatusBadRequest, "The email address is already associated with another user.")
-			return
+		switch err.(type) {
```
Comment (arkadiuszos4chain) [inline transports/http/endpoints/api/users/endpoints.go:72]: It seems in all cases (beside `default`) we could use only one type of error, e.g. UserError with descriptive message from the throwing point (and as I can see you already add good decriptions for paymail and xpub errors).
Right now we will end up with messages like : `Error registering Paymail: error registering paymail in BUX: (internal error)`

File: transports/http/server/http_server_test.go
Comment (arkadiuszos4chain) [inline transports/http/server/http_server_test.go:1]: <3

File: transports/http/endpoints/api/users/endpoints.go (hunk 3 above)
Comment (ac4ch) [reply transports/http/endpoints/api/users/endpoints.go:72]: not really, this methods can throw 3 types of errors UserError PaymailError and XPubError as per the kind of api calls is doing .. and in switch u can see iam checking each type of the error instead comparing string

**Pull request merged**